### PR TITLE
feat: swagger io

### PR DESCRIPTION
**Pull request by ascariandrea**
🎉  API definition with Swagger in Settings 🎉  

Screenshot from 2021-11-18 18-28-04

The "models" definitions are still missing, but they can be easily added - maybe in the next week.

### CODE REVIEW

**Review by djfm — COMMENTED** [review]
Looks cool!

But I think there are a few issues, for instance this route does not exist (at least, is not a POST):
image

And the routes that require the `x-authorization` header are pretty hard to use, it would be nice if the field could be pre-populated with the current `x-authorization` header when you try the endpoints out.

Lastly, it seems to send request to the production server even in development mode, is that by design?

**Comment by vecna** [comment]
This is supercool and I love it, but one minor question and a blocking problem:

minor: why you need the "execute" button? it is necessary to enable field editing (e.g. accessToken or other parameters) so it is not really an execution, rather an "open"

major: many API aren't working (for example, compare or author) how is the list built so I can review it? 

thanks!

**Comment by ascariandrea** [comment]
@vecna consider the UI is given by Swagger and we can't customize it easily, but:
- the execute button let you trigge the request by filling the params directly in the UI
- you can set the authorization token directly in requests, but if you click the "Authorize" blue button you can set the token for all authorized requests

File: src/providers/swagger/IOTSToOpenAPISchema.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
+/* eslint-disable no-case-declarations */
```
Comment (ascariandrea) [inline src/providers/swagger/IOTSToOpenAPISchema.ts:1]: this files tranform io-ts codecs into Open API schema (@vecna )

File: src/providers/swagger/swagger.provider.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
+/**
```
Comment (ascariandrea) [inline src/providers/swagger/swagger.provider.ts:1]: here's the file where we parse and transform the endpoints in proper swagger configuration

File: src/providers/swagger/swagger.provider.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    security,
+    requestBody,
+    responses: {
+      ...successResponse,
+    },
+  };
+};
+
+export const generateDoc = (config: DocConfig): any => {
+  const paths = pipe(
```
Comment (ascariandrea) [inline src/providers/swagger/swagger.provider.ts:183]: So basically this walks though endpoints exported from backend and generate an object with paths

Comment (ascariandrea) [reply src/providers/swagger/swagger.provider.ts:183]: @vecna

File: src/providers/swagger/swagger.provider.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+              in: 'header',
+              required: false,
+              schema: getOpenAPISchema((Headers?.props)[k]),
+            }),
+          []
+        )
+      : [];
+
+  // add security when "x-authorization" header is defined
+  const security =
```
Comment (ascariandrea) [inline src/providers/swagger/swagger.provider.ts:98]: This line enables authentication if the request has `x-authorization` token

File: src/providers/swagger/swagger.provider.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+  // derive headers from io-ts definition of e.Input.Headers
+  const headerParameters =
+    Headers !== undefined
+      ? R.keys(Headers.props).reduce<any[]>(
+          (acc, k) =>
+            acc.concat({
+              name: k,
+              in: 'header',
+              required: false,
```
Comment (ascariandrea) [inline src/providers/swagger/swagger.provider.ts:90]: headers should be required anyway

File: src/providers/swagger/swagger.provider.ts [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+              );
+              const prevEndpoints =
+                (endpointAcc as any)[endpointStaticPath] ?? undefined;
+
+              const previousMethodSchema =
+                prevEndpoints?.[endpoint.Method.toLowerCase()];
+              const currentEndpointSchema = apiSchemaFromEndpoint(
+                key,
+                endpoint,
+                [`${versionKey} - ${scopeKey}`]
```
Comment (ascariandrea) [inline src/providers/swagger/swagger.provider.ts:204]: this is the label for grouping the api by `version` and `scope` (public/creator)

**Comment by ascariandrea** [comment]
> Looks cool!
> 
> But I think there are a few issues, for instance this route does not exist (at least, is not a POST): image
> 

This is actually the correct definition. 

I changed the implementation cause it's wrong https://github.com/tracking-exposed/yttrex/pull/136

**Comment by ascariandrea** [comment]
we're merging this before moving everything inside the yttrex repo. We can iterate on this later.

**Pull request merged**